presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java

```diff
@@ -331,20 +343,20 @@ private FunctionHandle lookupFunction(
         throw new PrestoException(FUNCTION_NOT_FOUND, constructFunctionNotFoundErrorMessage(functionName.toString(), parameterTypes, candidates));
     }
 
-    private Optional<FunctionNamespaceManager<?>> getServingFunctionNamespaceManager(QualifiedFunctionName functionName)
+    private Optional<FunctionNamespaceManager<?>> getServingFunctionNamespaceManager(CatalogSchemaName functionNamespace)
     {
-        QualifiedFunctionName.Prefix functionPrefix = functionName.getPrefix();
-        if (functionPrefix.equals(DEFAULT_NAMESPACE)) {
+        if (functionNamespace.equals(DEFAULT_NAMESPACE)) {
             return Optional.of(builtInFunctionNamespaceManager);
         }
 
-        QualifiedFunctionName.Prefix bestMatchNamespace = null;
+        CatalogSchemaPrefix bestMatch = null;
         FunctionNamespaceManager<?> servingFunctionNamespaceManager = null;
 
-        for (Map.Entry<QualifiedFunctionName.Prefix, FunctionNamespaceManager<?>> functionNamespace : functionNamespaces.entrySet()) {
-            if (functionNamespace.getKey().contains(functionPrefix) && (bestMatchNamespace == null || bestMatchNamespace.contains(functionNamespace.getKey()))) {
-                bestMatchNamespace = functionNamespace.getKey();
-                servingFunctionNamespaceManager = functionNamespace.getValue();
+        for (Map.Entry<CatalogSchemaPrefix, FunctionNamespaceManager<?>> entry : functionNamespaces.entrySet()) {
+            CatalogSchemaPrefix prefix = entry.getKey();
+            if (prefix.includes(functionNamespace) && (bestMatch == null || bestMatch.includes(prefix))) {
```

Comment:
Maybe have a `includesButNotEqual` instead. Currently this is relying on same prefix would not appear more than once.

Reply:
I'm not sure what will be the behavior of the `includesButNotEqual` method you mentioned here.

Reply:
the `includes` method in prefix returns true of it has a bigger or equal scope. I'm suggesting a method that would just return true if it actually has a bigger scope. For example `"a.b" includes "a.b" => true`, `"a.b" includes but not equal "a.b" => false`.